src/fluka/fluka_funcs.cpp

```diff
@@ -0,0 +1,1104 @@
+  if (moab::MB_SUCCESS != rval)
+    fludag_abort("f_normal", "DAGMC failed in test_volume_boundary", rval);
+
+  rval = DAG->get_angle(state.next_surface, xyz, norml);
+  if (moab::MB_SUCCESS != rval)
+    fludag_abort("f_normal", "DAGMC failed in get_angle", rval);
+
+  // result = 1 entering, 0 leaving
+  if (result == 0) {   // vector should point towards OldReg
+    norml[0] = norml[0] * -1.0;
+    norml[1] = norml[1] * -1.0;
+    norml[2] = norml[2] * -1.0;
+  }
+
+  if (debug)
+    std::cout << "Normal: " << norml[0] << ", " << norml[1] << ", " << norml[2]  << std::endl;
+
+  return;
+}
+
+/* does the position pos belong to region oldRegion */
+inline bool check_vol(double pos[3], double dir[3], int oldRegion) {
+  int is_inside; // in volume or not
+  // convert region id into entityhandle
+  moab::EntityHandle volume = DAG->entity_by_index(3, oldRegion); // get the volume by index
+  moab::ErrorCode rval = DAG->point_in_volume(volume, pos, is_inside, dir);
+
+  // check for non error
+  if (moab::MB_SUCCESS != rval)
+    fludag_abort("check_vol", "DAGMC failed in point_in_volume", rval);
+
+  if (is_inside == 1)   // we are inside the cell tested
+    return true;
+  else
+    return false;
+}
+
+
+/* testable wrapper function for f_look */
+int look(double& posx, double& posy, double& posz, double* dir, int& oldRegion) {
+  int flagErr;
+  int lattice_dummy;  // not used
+  int nextRegion;
+  f_look(posx, posy, posz, dir, oldRegion, lattice_dummy, nextRegion, flagErr, lattice_dummy);
+  return nextRegion;
+}
+
+
+/* determine where a particle is, given position, direction etc */
+void f_look(double& pSx, double& pSy, double& pSz,
+            double* pV, const int& oldReg, const int& oldLttc,
+            int& nextRegion, int& flagErr, int& newLttc) {
+  if (debug) {
+    std::cout << "-------f_look------" << std::endl;
+    std::cout << " stack size is " << flkstk_.npflka << std::endl;
+    std::cout << "position is " << pSx << " " << pSy << " " << pSz << std::endl;
+  }
+
+  reset_state(state);
+  // get the stack count
+  //  state.stack_count = flkstk_.npflka;
+
+  const double xyz[] = {pSx, pSy, pSz};       // location of the particle (xyz)
+  const double dir[] = {pV[0], pV[1], pV[2]};
+
+  int is_inside = 0;
+  int num_vols = DAG->num_entities(3);  // number of volumes
+
+  for (int i = 1 ; i <= num_vols ; i++) { // loop over all volumes
+    moab::EntityHandle volume = DAG->entity_by_index(3, i); // get the volume by index
+    // No ray history  - doesnt matter, only called for new source particles
+    moab::ErrorCode rval = DAG->point_in_volume(volume, xyz, is_inside, dir);
+    // check for non error
+    if (moab::MB_SUCCESS != rval)
+      fludag_abort("f_look", "DAGMC failed in point_in_volume", rval);
+
+    // we think that we are inside ,
+    if (is_inside == 1) {   // we are inside the cell tested
+      // test point in vol again, but reverse the direction
+      double new_dir[3];
+      new_dir[0] = -1.*dir[0];
```

Comment:
Here and throughout, there are a lot of points in the code where the three axis of a point are managed via a `double[3]`.

May be helpful and more readable to create a `struct` or helper class to manage and encapsulate the point / linear algebra math logic.  This class can also provide special accessor and ingestors to interact with `DAG`.

Encapsulating this logic can lead to less future bugs and easier readability, as well as making the code less intimidating to newcomers overall.